include/bluetooth/audio/tbs.h

```diff
@@ -0,0 +1,331 @@
+#define BT_TBS_TECHNOLOGY_2G                       0x08
+#define BT_TBS_TECHNOLOGY_WCDMA                    0x09
+#define BT_TBS_TECHNOLOGY_IP                       0x0a
+
+/**
+ * @brief The GTBS index denotes whenever a callback is from a
+ * Generic Telephone Bearer Service (GTBS) instance, or
+ * whenever the client should perform on action on the GTBS instance of the
+ * server, rather than any of the specific Telephone Bearer Service instances.
+ */
+#define BT_TBS_GTBS_INDEX                               0xFF
+
+/**
+ * @brief Callback function for client originating a call.
+ *
+ * @param conn          The connection used.
+ * @param call_index    The call index.
+ * @param uri           The URI. The value may change, so should be
+ *                      copied if persistence is wanted.
+ *
+ * @return true if the call request was accepted and remote party is alerted.
+ */
+typedef bool (*bt_tbs_originate_call_cb)(struct bt_conn *conn,
+					 uint8_t call_index,
+					 const char *uri);
+
+/**
+ * @brief Callback function for client terminating a call.
+ *
+ * The call may be either terminated by the client or the server.
+ *
+ * @param conn          The connection used.
+ * @param call_index    The call index.
+ * @param reason        The termination BT_TBS_REASON_* reason.
+ */
+typedef void (*bt_tbs_terminate_call_cb)(struct bt_conn *conn,
+					 uint8_t call_index,
+					 uint8_t reason);
+
+/**
+ * @brief Callback function for client joining calls.
+ *
+ * @param conn                  The connection used.
+ * @param call_index_count      The number of call indexes to join.
+ * @param call_indexes          The call indexes.
+ */
+typedef void (*bt_tbs_join_calls_cb)(struct bt_conn *conn,
+				     uint8_t call_index_count,
+				     const uint8_t *call_indexes);
+
+/**
+ * @brief Callback function for client request call state change
+ *
+ * @param conn          The connection used.
+ * @param call_index    The call index.
+ */
+typedef void (*bt_tbs_call_change_cb)(struct bt_conn *conn,
+				      uint8_t call_index);
+
+/**
+ * @brief Callback function for authorizing a client.
+ *
+ * Only used if BT_TBS_AUTHORIZATION is enabled.
+ *
+ * @param conn         The connection used.
+ *
+ * @return true if authorized, false otherwise
+ */
+typedef bool (*bt_tbs_authorize_cb)(struct bt_conn *conn);
+
+struct bt_tbs_cb {
+	bt_tbs_originate_call_cb      originate_call;
+	bt_tbs_terminate_call_cb      terminate_call;
+	bt_tbs_call_change_cb         hold_call;
+	bt_tbs_call_change_cb         accept_call;
+	bt_tbs_call_change_cb         retrieve_call;
+	bt_tbs_join_calls_cb          join_calls;
+	bt_tbs_authorize_cb           authorize;
+};
+
+/**
+ * @brief Accept an alerting call.
+ *
+ * @param call_index    The index of the call that will be accepted.
+ *
+ * @return int          BT_TBS_RESULT_CODE_* if positive or 0,
+ *                      errno value if negative.
+ */
+int bt_tbs_accept(uint8_t call_index);
```

Comment:
If I were telephone service implementer that would expose the functionality over this API that would be a difficult for me TBH. When I should call this function (and related)? I have mixed feelings because it looks like it shall be called to accept the call as the name suggests.

If that's Server side API why this function isn't called e.g. *bt_tbs_call_accepted* ?
I guess the intention is not to accept the call by calling this function but rather indicate that the call has been accepted by a service that implements telephone functionality.

Or API user should perform the call acceptance from accept_call callback context? I guess not, because it would be odd that Bluetooth defines API for managing telephone service

Reply:
This is a server side API indeed. `bt_tbs_call_accepted` is a better name, and will update :) 

We don't want to use a callback for this, as the time between the server putting a call in the calling state, and when the call is actually accepted can be very long.

Reply:
> This is a server side API indeed. `bt_tbs_call_accepted` is a better name, and could be updated :) However, if we update this, then `bt_tbs_hold`, `bt_tbs_retrieve` etc. should similar be updated. Should we rename all the server APIs for controlling the call state be of the pattern `bt_tbs_call_Xed` instead of bt `bt_tbs_X`?

I do not want to push, but that's my feeling.
The Server API user by calling those functions wants to notify the call state changed (e.g. was accepted) not to change the call state (e.g. to accept).

Just like:
bt_conn_disconnect <- initiate termination
bt_conn_disconnected <- already terminated, tell anyone
^^

> 
> We don't want to use a callback for this, as the time between the server putting a call in the calling state, and when the call is actually accepted can be very long.
Understood, I wanted to show this can be misused.

Reply:
> The Server API user by calling those functions wants to notify the call state changed (e.g. was accepted) not to change the call state (e.g. to accept).

The calls and the states reside in the TBS instance at this moment, so these function do change the state and do not just notify the state changed.

Reply:
But the API user does not know the TBS implementation internals, does not have to

Reply:
> But the API user does not know the TBS implementation internals, does not have to

I'm sorry, I don't think I understand what you mean by this. Can you elaborate?

Reply:
> The calls and the states reside in the TBS instance at this moment, so these function do change the state and do not just notify the state changed.

I mean the API user - e.g. telephone implementation will most likely hold the calls and it's states is it's own structures. API user does not have to know whether the TBS holds it's copy or not, it does not matter. So the function name shall indicate that API user should call this function when the call state changed - because this function will notify about this change. The way how it's implemented (it stores the copy of the state as you said) does not matter.

Reply:
> I mean the API user - e.g. telephone implementation will most likely hold the calls and it's states is it's own structures. 

That is not how TBS is currently implemented, and based on the discussions in the F2F it also haven't been decided whether that should be the case or not yet.

Reply:
I think it's natural that methods have names in verb-form, and "bt_tbs_call_accepted" can be parsed in verb-form as "call the accepted thing", which is wrong. Whatever we go for, we should minimize ambiguity. I prefer "bt_tbs_accept" or "bt_tbs_set_call_accepted".